pkg/supportconfig/auth.go

```diff
@@ -0,0 +1,140 @@
+package supportconfig
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	authv1 "k8s.io/api/authentication/v1"
+	v1 "k8s.io/api/authorization/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	authnv1 "k8s.io/client-go/kubernetes/typed/authentication/v1"
+	authzv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+)
+
+const (
+	unauthenticatedError = "unauthenticated \n"
+	authHeader           = "Authorization"
+)
+
+type AuthHandler struct {
+	subjectReview authzv1.SubjectAccessReviewInterface
+	tokenReview   authnv1.TokenReviewInterface
+}
+
+// NewAuthHandler is an Auth handler for specifically the supportconfig endpoint
+func NewAuthHandler(sar authzv1.SubjectAccessReviewInterface, tr authnv1.TokenReviewInterface) *AuthHandler {
+	auth := &AuthHandler{
+		subjectReview: sar,
+		tokenReview:   tr,
+	}
+	return auth
+}
+
+func (a *AuthHandler) Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		userInfo, err := a.authenticate(r)
+		if err != nil {
+			w.WriteHeader(http.StatusUnauthorized)
+			if err.Error() != unauthenticatedError {
+				logrus.Errorf("error authenticating user %s", err)
+			} else {
+				_, err = w.Write([]byte(err.Error()))
+				if err != nil {
+					logrus.Errorf("error writing response string in auth %s", err.Error())
+				}
+			}
+			// don't go any further if we got an error during auth
+			return
+		}
+		err = a.authorize(r, userInfo)
+		if err != nil {
+			w.WriteHeader(http.StatusForbidden)
+			_, err = w.Write([]byte(err.Error()))
+			if err != nil {
+				logrus.Errorf("error writing response string in auth %s", err.Error())
+			}
+			return
+		}
+		// if we didn't get any errors while authing, it's safe to call the next function
+		next.ServeHTTP(w, r)
+	})
+}
+
+// authenticate issues a token review for the K8s api - basically you need to have `Authorization: Bearer` set like you
+// would for the K8s api. Right now we only have one endpoint so doing authn and authz at the same time makes sense.
+// however, if this changes, this should be broken out into its own middleware
+func (a *AuthHandler) authenticate(r *http.Request) (*authv1.UserInfo, error) {
+	token, err := getTokenFromHeader(r)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := a.tokenReview.Create(r.Context(), &authv1.TokenReview{
+		Spec: authv1.TokenReviewSpec{
+			Token: token,
+		},
+	}, metav1.CreateOptions{})
```

Comment:
I don't fully understand the context of how and where this application will be run so it's hard to evaluate this, but it looks to me like this is mainly useful for authenticating service accounts directly with the kubernetes cluster. Is that how this is meant to be used? Are real users supposed to be able to generate API tokens to use here? Does the dashboard need to be able to use this?

Reply:
Apologies for the lack of context.

This is designed to run within EKS clusters running rancher 2.6.6 or higher. The functionality is to allow admin users to call this endpoint and generate a support bundle. In order to not re-implement significant auth functionality, I figured that we could just require users to supply the creds to this endpoint the same way they do to any other part of rancher. Since you can use tokens as kubeconfig tokens (and that's what the UI uses to get resources from rancher), I figured that the end result would be a link in the UI which redirects you to this service, and you supply the creds the same way you do to the rest of rancher. 

TL;DR - it's supposed to be used for getting a user from a supplied rancher api token (generated by the UI or user). This wasn't tested yet (I used an SA token when testing). Given your comment here, probably something that I should look at. Will post back here if it works.

Reply:
> Is the EKS cluster the local cluster or the downstream cluster?

Local cluster. 

> Yeah I couldn't get that to work, that token is a Rancher construct and doesn't go to k8s for validation

Ah. I should have tested that scenario, sorry about that. One thing to keep in mind - this won't go directly from the user/browser to the service. The request will need to be proxied through rancher (i.e. /k8s/local/ was how it was described to me). I'm thinking that due to impersonation the end creds that will be used will be an SA token, but that's just a guess. I'll test this out and get a better approach together that is more realistic.

Reply:
Ok - ran some tests and have some more info. 

The idea is that when the user clicks on "give me a supportconfig" button in the UI, it will redirect them to the service proxy URL for the service (`$RANCHER_URL/api/v1/namespaces/cattle-csp-system/services/https:rancher-csp-adapter:https/proxy/generate/supportconfig`). This does not appear to pass along the header like I thought. I will see what options I have here - maybe there is a header that I can use to do what I need.

Reply:
Resolving - supportconfig functionality will not be altered here at this time.